Records/Records.md

```diff
@@ -0,0 +1,202 @@
+two instances of the same record type (e.g. two persons).
+
+A record in Proton is immutable. That is, once it has been declared
+there is no way to mutate the data held in an instance of said record.
+Ways of dealing mutating values stored in a record will be presented
+though these are subject to change based on future proposals (specifically
+closures from @lalex).
+
+### Assumptions
+
+I am making the following assumptions:
+
+* Type names start with a capital letter (e.g. `Person` instead of `person`)
+* Variables and identifiers start with a lowercased letter
+* The type goes on the right side of an expression. (`age: Int` instead of `Int age`)
+* Instantiation does not require `new`
+ 
+Some of these assumptions might prove wrong in which case I will update
+this list accordingly. Also some might prove to be an inspiration for
+future proposals.
+
+### Features
+
+Just for clarity, the features outlined in this proposal is the following:
+
+1. Declaration of a record
+2. Mutation of an immutable record (though this may be dropped)
+
+Proposal
+--------
+
+### Motivation
+
+The driving motivation for records is to give a name to a collection of
+related values.
+It is important to notice that records is not the same as tuples.
+Whereas a tuple is a collection of values that happen to "travel"
+together, a record is a collection of values which together have
+semantic meaning.
+
+### Declaration
+
+Declaring a record should ideally be as simple as possible.
+I propose to introduce `record` as a keyword.
+Like in the following example:
+
+    record Person(name: String, age: Int)
+
+The part `(name: String, age: Int)` is referred to as the _record constructor_.
+Declaring a record would generate a function with the name of the record
+taking the number of parameters declared in the record constructor.
+
+Declaring a record would also generate accessors for each given member
+declared in the record constructor.
+As in the example with `Person` both `name` and `age` would be members of
+the `Person` record and accessors would be created for both members.
+
+### Instantiation
+
+Instantiating a record is done via the function generated when the record
+was declared.
+Thus to create an instance of the `Person` record, we would use the
+following form:
+
+    Person("John Doe", 99)
+
+#### Member Access
+
+Accessing a member of a record would be possible using dot-syntax.
+That is `record.memberName`.
+
+As the member of a record might be a collection or some other type
+that has members itself, this member access is going to be an expression
+such that further "dotting" is possible.
+
+### Mutation
+
+**NOTE:** This might need to go in its own separate proposal.
+
+Since records in this proposal is immutable we can't directly
+mutate values.
+We need some way to mutate and update values in a record.
+
+I propose to introduce the keyword `with` for mutating one or several
+members of a record.
```

Comment:
From my understanding, your want to keep records immutable by making a *copy* of the original record and then passing the new value for a given constructor parameter ([this is the general idea of how the `with` syntax works in F#][1]).

For example:
```
let p1 = Person("Jane Doe", 10)
let p2 = p1 with { age = 42 }
```
Could compile down to the following (assuming strings are immutable) to keep records immutable:
```
let p1 = Person("Jane Doe", 10)
let p2 = Person(p1.name, 42)
```

(I could always look at the IL that F# generates if you want to get overly technical here...)

I would update the wording to make it clear that records are *immutable* and that you are making a *copy* of the original record. I would probably use the word *update* instead of *mutate*. *Mutate* is more specific and general means you're modifying a value in-place. I think *update* seems a bit more general term and does not imply it's necessarily an in-place update.

I would suggest something along the lines of the following (you can probably mix this with your proposal better):

> I propose to introduce the keyword `with` to allow for updating one or several members of an immutable record. It serves as "syntactic sugar" for updating an existing record by making a copy of the original with updated member values.

You could even go as far as say that "a new record is instantiated by passing the original and the updated member values to their respective parameter in the record constructor".

It sounds nice and technical, but I don't know if that is necessary at this point since there is quite a lot still to be decided. Leaving some generalities open for later may be more appropriate.

 [1]: https://msdn.microsoft.com/en-us/visualfsharpdocs/conceptual/records-%5bfsharp%5d